File: apps/ARIA_MCP/Sources/AriaMCP/RecipeTools.swift
// RecipeTools.swift
//
// The CognitionKit behaviour-recipe surface on ARIA_MCP. These tools sit
// ABOVE the lexicon projection (provenance `.recipe`, dispatched by name)
// exactly as the federation `cross_estate_recall` tool does. They are the
// conscious-mind surface: the MCP↔CognitionKit channel is R/W, so an
// agent reads what recipes exist and triggers them, and the human-in-the-
// loop confirms the migration promotion.
//
// Four foundational tools ship here:
//   - moot_list_lenses           → ProjectedTool descriptor enumeration
//                                   (LensTools + RecipeTools, Tier 6 only — read)
//   - moot_synthesize            → GroundedSynthesis recipe (read)
//   - moot_run_migration         → MigrationBenchmark.run (read; no
//                                   promotion — B-3)
//   - moot_confirm_migration     → MigrationBenchmark.confirmPromotion
//                                   by branch id (the human-gated write)
//
// The analytics recipes (moot_association_rules, moot_formal_concepts) and
// all 14 reasoning-lens recipes ship through LensTools.swift — same
// provenance (.recipe), dispatched through the lens surface per
// LENS_DISCOVERABILITY_DECISION v2.0.
//
// Stateless-boundary discipline: MCP `tools/call` is stateless across
// invocations, so the run tool cannot hand live `BranchHandle`s to a
// later confirm tool. The run tool surfaces branch *ids* in its result;
// the confirm tool takes those ids and re-resolves them through the
// long-lived kit's `GeniusLocusKit.branchHandle(for:)` accessor. This is
// the two-call run→confirm pattern the plan describes; webhook delivery
// of async completion is an additive transport, not needed for the
// synchronous behaviour.

import Foundation
import GeniusLocusKit
import NeuronKit
import LocusKit
import CognitionKit

/// Namespace for the CognitionKit recipe tool surface. No instances.
enum RecipeTools {

    // MARK: - Tool names

    static let listRecipesToolName = "moot_list_lenses"
    static let groundedSynthesisToolName = "moot_synthesize"
    static let runMigrationBenchmarkToolName = "moot_run_migration"
    static let confirmMigrationPromotionToolName = "moot_confirm_migration"

    /// True when `name` is one of the foundational recipe tools dispatched by name.
    static func isRecipeTool(_ name: String) -> Bool {
        name == listRecipesToolName
            || name == groundedSynthesisToolName
            || name == runMigrationBenchmarkToolName
            || name == confirmMigrationPromotionToolName
    }

    // MARK: - tools/list projection

    /// All foundational recipe tool descriptors, advertised in `tools/list`
    /// after the lexicon projection and the federation tool. The analytics
    /// and reasoning-lens tools are projected by LensTools.
    static func tools() -> [ProjectedTool] {
        [
            listRecipesTool(),
            groundedSynthesisTool(),
            runMigrationBenchmarkTool(),
            confirmMigrationPromotionTool(),
        ]
    }

    /// The cognition-discovery tool. At runtime (`runListRecipes`), returns
    /// one block per Tier 6 tool drawn from `LensTools.tools()` and the two
    /// non-migration entries in `RecipeTools.tools()` — not `RecipeCatalog`.
    /// Takes no arguments; it is the conscious mind enumerating its surface.
    private static func listRecipesTool() -> ProjectedTool {
        ProjectedTool(
            name: listRecipesToolName,
            description: "List the available reasoning lenses and CognitionKit behaviour recipes — each with its version, description, and the NeuronKit capabilities it requires.",
            inputSchema: objectSchema(properties: [:], required: []),
            provenance: .recipe)
    }

    private static func groundedSynthesisTool() -> ProjectedTool {
        ProjectedTool(
            name: groundedSynthesisToolName,
            description: "Synthesize memories into a grounded context document: hybrid-recall and summarise into patterns, success rate, recommendations, and key insights.",
            inputSchema: objectSchema(
                properties: [
                    "filter": stringSchema("Filter kind: unconfirmed (default), userConfirmed, exportable, contained, currentlyBelieve."),
                    "limit": integerSchema("Max drawers to recall."),
                    "estateID": stringSchema("Optional UUID of the open estate to target. Omit for the default estate."),
                ],
                required: []),
            provenance: .recipe)
    }

    private static func runMigrationBenchmarkTool() -> ProjectedTool {
        ProjectedTool(
            name: runMigrationBenchmarkToolName,
            description: "Run a migration benchmark: derive COW branches, populate from origin corpus, benchmark recall fidelity, and rank survivors. Returns branch ids for a separate confirm step. Never promotes automatically.",
            inputSchema: objectSchema(
                properties: [
                    "corpusName": stringSchema("Human-readable name for the origin corpus."),
                    "entries": .object([
                        "type": .string("array"),
                        "description": .string("Origin entries: objects with string fields id, content, and a string-array field tags."),
                        "items": objectSchema(
                            properties: [
                                "id": stringSchema("Stable source id."),
                                "content": stringSchema("Verbatim content."),
                                "tags": .object([
                                    "type": .string("array"),
                                    "items": .object(["type": .string("string")]),
                                ]),
                            ],
                            required: ["id", "content"]),
                    ]),
                    "plans": .object([
                        "type": .string("array"),
                        "description": .string("Candidate plans: objects with string fields name, room, latticeCode, embeddingModelID, and optional sensitivity (normal/elevated/restricted/secret)."),
                        "items": objectSchema(
                            properties: [
                                "name": stringSchema("Plan name (also the branch name)."),
                                "room": stringSchema("Room every migrated drawer is filed into."),
                                "latticeCode": stringSchema("UDC lattice code for every migrated drawer."),
                                "embeddingModelID": stringSchema("Embedding model id tagged on every drawer."),
                                "sensitivity": stringSchema("Optional sensitivity tier; default normal."),
                            ],
                            required: ["name", "room", "latticeCode", "embeddingModelID"]),
                    ]),
                    "estateID": stringSchema("Optional UUID of the open estate to target. Omit for the default estate."),
                ],
                required: ["corpusName", "entries", "plans"]),
            provenance: .recipe)
    }

    private static func confirmMigrationPromotionTool() -> ProjectedTool {
        ProjectedTool(
            name: confirmMigrationPromotionToolName,
            description: "Confirm a migration: promote the winning branch into the estate and discard the losers, by branch id. Refuses to promote any branch the run report disqualified.",
            inputSchema: objectSchema(
                properties: [
                    "winnerBranchID": stringSchema("UUID of the winning branch to promote (from the run report)."),
                    "discardBranchIDs": .object([
                        "type": .string("array"),
                        "description": .string("UUIDs of losing branches to discard."),
                        "items": .object(["type": .string("string")]),
                    ]),
                    "disqualifiedBranchIDs": .object([
                        "type": .string("array"),
                        "description": .string("UUIDs the run report disqualified; the C-5 guard refuses to promote any of these."),
                        "items": .object(["type": .string("string")]),
                    ]),
                    "estateID": stringSchema("Optional UUID of the open estate to target. Omit for the default estate."),
                ],
                required: ["winnerBranchID"]),
            provenance: .recipe)
    }

    // MARK: - Dispatch

    /// Run the named recipe tool. `resolveHandle` is the dispatcher's own
    /// estateID→handle resolver (absent estateID ⇒ default handle), passed
    /// in so recipe tools honour the same multi-estate routing as lexicon
    /// tools. Out-of-band faults throw `JSONRPCError`; recipe-level
    /// refusals come back as `errorResult` (isError == true) so the client
    /// keeps the call id, matching the lexicon-tool discipline.
    static func dispatch(
        name: String,
        args: [String: JSONValue],
        kit: GeniusLocusKit,
        defaultHandle: EstateHandle,
        resolveHandle: ([String: JSONValue]) throws -> EstateHandle
    ) async throws -> JSONValue {
        // Recipe discovery needs no estate; answer it before resolving a
        // handle so `moot_list_lenses` works even with no estate targeted.
        if name == listRecipesToolName {
            return runListRecipes()
        }
        let handle = try resolveHandle(args)
        switch name {
        case groundedSynthesisToolName:
            return try await runGroundedSynthesis(args, kit: kit, handle: handle)
        case runMigrationBenchmarkToolName:
            return try await runMigrationBenchmark(args, kit: kit, handle: handle)
        case confirmMigrationPromotionToolName:
            return try await runConfirmPromotion(args, kit: kit, handle: handle)
        default:
            throw JSONRPCError(
                code: JSONRPCErrorCode.methodNotFound,
                message: "Unknown recipe tool: \(name)")
        }
    }

    // MARK: - list_lenses (cognition menu)

    /// Return a one-block-per-tool cognition menu drawn from the shipped
    /// `ProjectedTool` descriptors. Shows name, description, and required
    /// args for each Tier 6 cognition tool — the 16 lens tools plus
    /// `moot_synthesize` and `moot_list_lenses` itself.
    /// Migration tools (Tier 7) are intentionally excluded here; they have
    /// their own teachme guides and a separate caller workflow.
    private static func runListRecipes() -> JSONValue {
        // Tier 6 recipe tools: list-lenses + synthesize only (not migration).
        let tier6RecipeNames: Set<String> = [listRecipesToolName, groundedSynthesisToolName]
        let recipeTools = tools().filter { tier6RecipeNames.contains($0.name) }
        // All 16 lens tools from LensTools.
        let lensTools = LensTools.tools()
        let cognitionTools = recipeTools + lensTools

        var lines: [String] = ["\(listRecipesToolName): \(cognitionTools.count) cognition tools"]
        for tool in cognitionTools {
            let required = requiredArgNames(from: tool.inputSchema)
            let requiredText = required.isEmpty ? "none" : required.joined(separator: ", ")
            lines.append("")
            lines.append(tool.name)
            lines.append("  \(tool.description)")
            lines.append("  Required: \(requiredText).")
        }
        lines.append("")
        lines.append("Call any tool with teachme:true for a full usage guide.")
        return ToolDispatcher.textResult(lines.joined(separator: "\n"))
    }

    /// Extract required argument names from a JSON Schema `required` array.
    private static func requiredArgNames(from schema: JSONValue) -> [String] {
        schema.objectValue?["required"]?.arrayValue?.compactMap { $0.stringValue } ?? []
    }

    // MARK: - grounded_synthesis

    private static func runGroundedSynthesis(
        _ args: [String: JSONValue],
        kit: GeniusLocusKit,
        handle: EstateHandle
    ) async throws -> JSONValue {
        let filter = decodeFilter(args["filter"]?.stringValue)
        let limit = args["limit"]?.integerValue.map(Int.init)
        let frame = LocusKit.RecallFrame(
            filterChain: [filter],
            hydrationLevel: .structured,
            limit: limit,
            ordering: .byCaptureTimeDesc)

        let out = try await GroundedSynthesis().run(
            input: .init(frame: frame), estate: handle, kit: kit)

        let doc = out.context
        let body = """
        grounded_synthesis: \(out.drawerCount) drawer(s)
        summary: \(doc.summary)
        patterns: \(doc.patterns.joined(separator: ", "))
        successRate: \(doc.successRate)
        recommendations:
        \(doc.recommendations.map { "  - \($0)" }.joined(separator: "\n"))
        keyInsights:
        \(doc.keyInsights.map { "  - \($0)" }.joined(separator: "\n"))
        """
        return ToolDispatcher.textResult(body)
    }

    // MARK: - run_migration_benchmark

    private static func runMigrationBenchmark(
        _ args: [String: JSONValue],
        kit: GeniusLocusKit,
        handle: EstateHandle
    ) async throws -> JSONValue {
        let corpusName = try requireString(args, "corpusName")
        let entries = try decodeEntries(args["entries"])
        let plans = try decodePlans(args["plans"])
        guard !plans.isEmpty else {
            throw JSONRPCError(
                code: JSONRPCErrorCode.invalidParams,
                message: "run_migration_benchmark requires at least one plan")
        }

        let origin = ExternalCorpus(name: corpusName, entries: entries)
        let out = try await MigrationBenchmark().run(
            input: .init(origin: origin, plans: plans),
            estate: handle, kit: kit)

        let report = out.comparisonReport
        var lines: [String] = ["run_migration_benchmark"]
        if let wid = report.winnerBranchID, let wname = report.winnerPlanName {
            lines.append("winner: plan '\(wname)' branch \(wid)")
        } else {
            lines.append("winner: none (all plans disqualified or no plans)")
        }
        lines.append("rankings:")
        for r in report.rankings {
            lines.append("  - \(r.planName) [\(r.branchID)] score=\(r.combinedScore) overlap=\(r.recallOverlap) mrr=\(r.meanReciprocalRank)")
        }
        lines.append("disqualified:")
        for d in report.disqualified {
            lines.append("  - \(d.planName) [\(d.branchID)] lost: \(d.lostConcepts.joined(separator: ", "))")
        }
        lines.append("")
        lines.append("To promote, call \(confirmMigrationPromotionToolName) with winnerBranchID, discardBranchIDs (the other ranking ids), and disqualifiedBranchIDs from above.")
        return ToolDispatcher.textResult(lines.joined(separator: "\n"))
    }

    // MARK: - confirm_migration_promotion

    private static func runConfirmPromotion(
        _ args: [String: JSONValue],
        kit: GeniusLocusKit,
        handle: EstateHandle
    ) async throws -> JSONValue {
        let winner = try requireUUID(args, "winnerBranchID")
        let discard = try decodeUUIDArray(args["discardBranchIDs"])
        let disqualified = Set(try decodeUUIDArray(args["disqualifiedBranchIDs"]))

        do {
            try await MigrationBenchmark().confirmPromotion(
                winnerBranchID: winner,
                discardBranchIDs: discard,
                disqualifiedBranchIDs: disqualified,
                estate: handle, kit: kit)
        } catch let error as RecipeError {
            // Recipe-level refusal (e.g. silentConceptLoss on a
            // disqualified winner, or an unknown branch). The call reached
            // the recipe and was refused; surface it as a tool error so
            // the client keeps the call id and sees why.
            return ToolDispatcher.errorResult(error.description)
        }
        return ToolDispatcher.textResult(
            "confirm_migration_promotion: promoted \(winner); discarded \(discard.count) branch(es).")
    }

    // MARK: - Argument decoding

    private static func requireString(
        _ args: [String: JSONValue], _ key: String
    ) throws -> String {
        guard let v = args[key]?.stringValue else {
            throw JSONRPCError(
                code: JSONRPCErrorCode.invalidParams,
                message: "Missing required string argument: \(key)")
        }
        return v
    }

    private static func requireUUID(
        _ args: [String: JSONValue], _ key: String
    ) throws -> UUID {
        let raw = try requireString(args, key)
        guard let uuid = UUID(uuidString: raw) else {
            throw JSONRPCError(
                code: JSONRPCErrorCode.invalidParams,
                message: "Argument \(key) is not a UUID: \(raw)")
        }
        return uuid
    }

    private static func decodeUUIDArray(_ value: JSONValue?) throws -> [UUID] {
        guard let arr = value?.arrayValue else { return [] }
        var out: [UUID] = []
        out.reserveCapacity(arr.count)
        for element in arr {
            guard let s = element.stringValue, let uuid = UUID(uuidString: s) else {
                throw JSONRPCError(
                    code: JSONRPCErrorCode.invalidParams,
                    message: "Branch id array contains a non-UUID value")
            }
            out.append(uuid)
        }
        return out
    }

    /// Decode the origin corpus entries. Each entry needs string `id` and
    /// `content`; `tags` is an optional string array.
    private static func decodeEntries(_ value: JSONValue?) throws -> [ExternalEntry] {
        guard let arr = value?.arrayValue else {
            throw JSONRPCError(
                code: JSONRPCErrorCode.invalidParams,
                message: "entries must be an array")
        }
        return try arr.map { element in
            guard let obj = element.objectValue,
                  let id = obj["id"]?.stringValue,
                  let content = obj["content"]?.stringValue else {
                throw JSONRPCError(
                    code: JSONRPCErrorCode.invalidParams,
                    message: "each entry needs string id and content")
            }
            let tags = (obj["tags"]?.arrayValue ?? []).compactMap { $0.stringValue }
            return ExternalEntry(id: id, content: content, tags: tags)
        }
    }

    /// Decode the candidate migration plans.
    private static func decodePlans(_ value: JSONValue?) throws -> [MigrationPlan] {
        guard let arr = value?.arrayValue else {
            throw JSONRPCError(
                code: JSONRPCErrorCode.invalidParams,
                message: "plans must be an array")
        }
        return try arr.map { element in
            guard let obj = element.objectValue,
                  let name = obj["name"]?.stringValue,
                  let room = obj["room"]?.stringValue,
                  let code = obj["latticeCode"]?.stringValue,
                  let model = obj["embeddingModelID"]?.stringValue else {
                throw JSONRPCError(
                    code: JSONRPCErrorCode.invalidParams,
                    message: "each plan needs name, room, latticeCode, embeddingModelID")
            }
            let sensitivity = decodeSensitivity(obj["sensitivity"]?.stringValue)
            return MigrationPlan(
                name: name, room: room, latticeCode: code,
                embeddingModelID: model, sensitivity: sensitivity)
        }
    }

    /// Recall filter decode for grounded_synthesis. Defaults to
    /// `.unconfirmed` so freshly-captured rows (provenance == 0) are
    /// visible — the default recall prepend would otherwise prune them
    /// behind `.userConfirmed`.
    private static func decodeFilter(_ name: String?) -> LocusKit.Filter {
        switch name {
        case "userConfirmed": return .userConfirmed
        case "exportable": return .exportable
        case "contained": return .contained
        case "currentlyBelieve": return .currentlyBelieve
        default: return .unconfirmed
        }
    }

    private static func decodeSensitivity(_ name: String?) -> AdjectiveSensitivity {
        switch name {
        case "elevated": return .elevated
        case "restricted": return .restricted
        case "secret": return .secret
        default: return .normal
        }
    }

    // MARK: - JSON schema helpers
    //
    // ToolProjection's schema helpers are private to that type, so the
    // recipe surface carries its own minimal copies. Identical shapes —
    // a future refactor could promote one shared set.

    private static func objectSchema(
        properties: [String: JSONValue], required: [String]
    ) -> JSONValue {
        var schema: [String: JSONValue] = [
            "type": .string("object"),
            "properties": .object(properties),
        ]
        if !required.isEmpty {
            schema["required"] = .array(required.map { .string($0) })
        }
        return .object(schema)
    }

    private static func stringSchema(_ description: String) -> JSONValue {
        .object(["type": .string("string"), "description": .string(description)])
    }

    private static func integerSchema(_ description: String) -> JSONValue {
        .object(["type": .string("integer"), "description": .string(description)])
    }
}

File: apps/ARIA_MCP/rust/src/recipe_tools.rs
//! CognitionKit recipe tool surface — moot_list_lenses, moot_synthesize,
//! moot_run_migration, moot_confirm_migration.
//!
//! Mirrors Swift `RecipeTools.swift`. Same dispatch contract: out-of-band faults
//! throw `JSONRPCError`; recipe-level refusals come back as `error_result` (isError
//! true) so the client keeps the call id.
//!
//! # moot_run_migration
//!
//! Derives one COW branch per candidate plan, populates each from the origin corpus,
//! benchmarks recall fidelity with the zero-silent-loss gate, and returns a ranked
//! list of survivors. Never promotes. Returns branch ids in the result text so a
//! caller can pass them to the confirm tool.
//!
//! # moot_confirm_migration
//!
//! Dispatches `confirm_migration_promotion_by_id` — the id-addressed form of the
//! human-gated promotion step. The server is stateless across tool calls; the
//! two-call pattern (run then confirm) works because the coordinator retains all
//! minted branches in memory, and the run result text carries the branch ids the
//! caller needs. Required: `winnerBranchID` (UUID string). Optional:
//! `discardBranchIDs` and `disqualifiedBranchIDs` (arrays of UUID strings). Mirrors
//! the Swift server's confirm handler contract and the Swift
//! `MigrationBenchmark.confirmPromotion` by-id overload as the behavioral reference.

use std::collections::BTreeMap;

use cognition_kit::{recipe_catalog, run_grounded_synthesis, OriginEntry, PlanInput};
use genius_locus_kit::branches::BranchId;
use locus_kit::{
    adjectives::AdjectiveSensitivity,
    filter::{Filter, HydrationLevel, Ordering, RecallFrame},
};
use neuron_kit::RecallFrameTuning;
use uuid::Uuid;

use crate::dispatch::{error_result, require_string, text_result};
use crate::estate_registry::EstateRegistry;
use crate::jsonrpc::{JSONRPCError, JSONRPCErrorCode, JsonValue};

/// Recipe tool names — mirrors Swift `RecipeTools` static constants.
const LIST_LENSES: &str = "moot_list_lenses";
const SYNTHESIZE: &str = "moot_synthesize";
const RUN_MIGRATION: &str = "moot_run_migration";
const CONFIRM_MIGRATION: &str = "moot_confirm_migration";

/// True when `name` is one of the recipe tools.
pub fn is_recipe_tool(name: &str) -> bool {
    matches!(
        name,
        LIST_LENSES | SYNTHESIZE | RUN_MIGRATION | CONFIRM_MIGRATION
    )
}

/// Dispatch a recipe tool call. Same contract as `dispatch_tool`.
pub fn dispatch(
    name: &str,
    args: &BTreeMap<String, JsonValue>,
    registry: &EstateRegistry,
) -> Result<serde_json::Value, JSONRPCError> {
    match name {
        LIST_LENSES => Ok(run_list_recipes()),
        SYNTHESIZE => run_grounded_synthesis_tool(args, registry),
        RUN_MIGRATION => run_migration_benchmark_tool(args, registry),
        CONFIRM_MIGRATION => run_confirm_promotion_tool(args, registry),
        _ => Err(JSONRPCError::new(
            JSONRPCErrorCode::METHOD_NOT_FOUND,
            format!("Unknown recipe tool: {name}"),
        )),
    }
}

// ---------------------------------------------------------------------------
// moot_list_lenses
// ---------------------------------------------------------------------------

fn run_list_recipes() -> serde_json::Value {
    let catalog = recipe_catalog();
    let mut lines = vec![format!("recipes: {}", catalog.len())];
    for d in &catalog {
        let caps: Vec<&str> = d
            .required_capabilities
            .iter()
            .map(|c| {
                // The capability rawValue — matches the Swift .rawValue string.
                // Delegates to the capability's own raw_value() to stay in sync
                // with declaration order and avoid exhaustive match maintenance.
                c.raw_value()
            })
            .collect();
        lines.push(format!("  - {} v{}", d.name, d.version));
        lines.push(format!("      {}", d.description));
        lines.push(format!("      capabilities: {}", caps.join(", ")));
    }
    text_result(&lines.join("\n"))
}

// ---------------------------------------------------------------------------
// moot_grounded_synthesis
// ---------------------------------------------------------------------------

fn run_grounded_synthesis_tool(
    args: &BTreeMap<String, JsonValue>,
    registry: &EstateRegistry,
) -> Result<serde_json::Value, JSONRPCError> {
    let estate = registry.resolve(args, "estateID")?;
    let filter = decode_recipe_filter(args);
    let limit = args
        .get("limit")
        .and_then(|v| v.as_i64())
        .map(|i| i as usize);

    let mut frame = RecallFrame::new(vec![filter]);
    frame.hydration_level = HydrationLevel::Structured;
    frame.ordering = Ordering::ByCaptureTimeDesc;
    if let Some(l) = limit {
        frame.limit = Some(l);
    }

    let now = crate::dispatch::wall_now();
    let coord = estate.coord.lock().unwrap();
    let out = run_grounded_synthesis(
        &coord,
        &estate.handle,
        frame,
        RecallFrameTuning::default(),
        now,
    )
    .map_err(error_from_recipe)?;

    let doc = &out.context;
    let body = format!(
        "grounded_synthesis: {} drawer(s)\nsummary: {}\npatterns: {}\nsuccessRate: {}\nrecommendations:\n{}\nkeyInsights:\n{}",
        out.drawer_count,
        doc.summary,
        doc.patterns.join(", "),
        doc.success_rate,
        doc.recommendations.iter().map(|r| format!("  - {r}")).collect::<Vec<_>>().join("\n"),
        doc.key_insights.iter().map(|i| format!("  - {i}")).collect::<Vec<_>>().join("\n"),
    );
    Ok(text_result(&body))
}

// ---------------------------------------------------------------------------
// moot_run_migration_benchmark
// ---------------------------------------------------------------------------

fn run_migration_benchmark_tool(
    args: &BTreeMap<String, JsonValue>,
    registry: &EstateRegistry,
) -> Result<serde_json::Value, JSONRPCError> {
    let estate = registry.resolve(args, "estateID")?;
    let corpus_name = require_string(args, "corpusName")?;
    let entries = decode_entries(args)?;
    let plans = decode_plans(args)?;
    if plans.is_empty() {
        return Err(JSONRPCError::new(
            JSONRPCErrorCode::INVALID_PARAMS,
            "run_migration_benchmark requires at least one plan",
        ));
    }
    let origin: Vec<OriginEntry> = entries
        .into_iter()
        .map(|(id, content)| OriginEntry { id, content })
        .collect();

    let now = crate::dispatch::wall_now();
    let mut coord = estate.coord.lock().unwrap();

    use cognition_kit::migration_live::LiveRecipeSubstrate;
    // EstateHandle is Copy — use directly without clone.
    let mut substrate = LiveRecipeSubstrate::new(&mut coord, estate.handle, now);

    let report = cognition_kit::run_migration_benchmark(&mut substrate, &plans, &origin)
        .map_err(error_from_recipe)?;

    let mut lines = vec![format!("run_migration_benchmark: corpus={corpus_name}")];
    if let Some(winner) = &report.winner {
        let bid = report
            .plan_results
            .iter()
            .find(|p| &p.name == winner)
            .map(|p| p.branch_id.clone())
            .unwrap_or_default();
        lines.push(format!("winner: plan '{winner}' branch {bid}"));
    } else {
        lines.push("winner: none (all plans disqualified or no plans)".to_owned());
    }
    lines.push("rankings:".to_owned());
    for r in &report.rankings {
        let bid = report
            .plan_results
            .iter()
            .find(|p| p.name == r.name)
            .map(|p| p.branch_id.clone())
            .unwrap_or_default();
        lines.push(format!(
            "  - {} [{}] score={:.4} overlap={:.4} mrr={:.4}",
            r.name, bid, r.combined_score, r.recall_overlap, r.mean_reciprocal_rank
        ));
    }
    lines.push("disqualified:".to_owned());
    for d in &report.disqualified {
        let bid = report
            .plan_results
            .iter()
            .find(|p| p.name == d.name)
            .map(|p| p.branch_id.clone())
            .unwrap_or_default();
        lines.push(format!(
            "  - {} [{}] lost: {}",
            d.name,
            bid,
            d.lost_concepts.join(", ")
        ));
    }
    lines.push(String::new());
    lines.push(format!(
        "To promote, call {} with winnerBranchID, discardBranchIDs (the other ranking ids), and disqualifiedBranchIDs from above.",
        "moot_confirm_migration"
    ));
    Ok(text_result(&lines.join("\n")))
}

// ---------------------------------------------------------------------------
// moot_confirm_migration_promotion
// ---------------------------------------------------------------------------

fn run_confirm_promotion_tool(
    args: &BTreeMap<String, JsonValue>,
    registry: &EstateRegistry,
) -> Result<serde_json::Value, JSONRPCError> {
    // Parse winnerBranchID — required UUID string. Missing or malformed →
    // invalidParams (transport-level fault, mirrors server's existing pattern).
    let winner_str = require_string(args, "winnerBranchID")?;
    let winner_bid: BranchId = Uuid::parse_str(winner_str).map_err(|_| {
        JSONRPCError::new(
            JSONRPCErrorCode::INVALID_PARAMS,
            format!("Malformed winnerBranchID (not a UUID): {winner_str}"),
        )
    })?;

    // Parse optional discardBranchIDs — array of UUID strings; malformed
    // element → invalidParams.
    let discard_bids = parse_uuid_array(args, "discardBranchIDs")?;

    // Parse optional disqualifiedBranchIDs — same shape.
    let disqualified_bids = parse_uuid_array(args, "disqualifiedBranchIDs")?;

    // Resolve estate — absent estateID → default estate (v1 single-estate path).
    let estate = registry.resolve(args, "estateID")?;
    let now = crate::dispatch::wall_now();
    let mut coord = estate.coord.lock().unwrap();

    // Dispatch the by-id overload. RecipeError results surface as isError:true
    // tool results (lens-refusal discipline: call id kept). Substrate errors
    // surface as tool errors too — the coordinator failure detail is informative.
    match cognition_kit::migration_live::confirm_migration_promotion_by_id(
        &mut coord,
        winner_bid,
        &discard_bids,
        &disqualified_bids,
        &estate.handle,
        now,
    ) {
        Ok(()) => {
            // Mirror the Swift server's success text shape.
            let body = format!(
                "confirm_migration: promoted {}; discarded {} branch(es).",
                winner_str,
                discard_bids
                    .iter()
                    .filter(|&&bid| bid != winner_bid)
                    .count(),
            );
            Ok(text_result(&body))
        }
        // RecipeError results (SilentConceptLoss, UserConfirmationRequired) and
        // substrate errors surface as isError:true so the client keeps the call id.
        Err(e) => Ok(error_result(&format!("{e}"))),
    }
}

/// Parse an optional array of UUID strings from `args[key]`.
/// Absent key → empty vec. Malformed element → invalidParams.
fn parse_uuid_array(
    args: &BTreeMap<String, JsonValue>,
    key: &str,
) -> Result<Vec<BranchId>, JSONRPCError> {
    let Some(val) = args.get(key) else {
        return Ok(vec![]);
    };
    let arr = val.as_array().ok_or_else(|| {
        JSONRPCError::new(
            JSONRPCErrorCode::INVALID_PARAMS,
            format!("{key} must be an array of UUID strings"),
        )
    })?;
    arr.iter()
        .map(|v| {
            let s = v.as_str().ok_or_else(|| {
                JSONRPCError::new(
                    JSONRPCErrorCode::INVALID_PARAMS,
                    format!("each element of {key} must be a UUID string"),
                )
            })?;
            Uuid::parse_str(s).map_err(|_| {
                JSONRPCError::new(
                    JSONRPCErrorCode::INVALID_PARAMS,
                    format!("Malformed UUID in {key}: {s}"),
                )
            })
        })
        .collect()
}

// ---------------------------------------------------------------------------
// Argument decoders
// ---------------------------------------------------------------------------

fn decode_entries(
    args: &BTreeMap<String, JsonValue>,
) -> Result<Vec<(String, String)>, JSONRPCError> {
    let arr = args
        .get("entries")
        .and_then(|v| v.as_array())
        .ok_or_else(|| {
            JSONRPCError::new(JSONRPCErrorCode::INVALID_PARAMS, "entries must be an array")
        })?;
    arr.iter()
        .map(|element| {
            let obj = element.as_object().ok_or_else(|| {
                JSONRPCError::new(
                    JSONRPCErrorCode::INVALID_PARAMS,
                    "each entry must be an object",
                )
            })?;
            let id = obj.get("id").and_then(|v| v.as_str()).ok_or_else(|| {
                JSONRPCError::new(
                    JSONRPCErrorCode::INVALID_PARAMS,
                    "each entry needs string id",
                )
            })?;
            let content = obj.get("content").and_then(|v| v.as_str()).ok_or_else(|| {
                JSONRPCError::new(
                    JSONRPCErrorCode::INVALID_PARAMS,
                    "each entry needs string content",
                )
            })?;
            Ok((id.to_owned(), content.to_owned()))
        })
        .collect()
}

fn decode_plans(args: &BTreeMap<String, JsonValue>) -> Result<Vec<PlanInput>, JSONRPCError> {
    let arr = args
        .get("plans")
        .and_then(|v| v.as_array())
        .ok_or_else(|| {
            JSONRPCError::new(JSONRPCErrorCode::INVALID_PARAMS, "plans must be an array")
        })?;
    arr.iter()
        .map(|element| {
            let obj = element.as_object().ok_or_else(|| {
                JSONRPCError::new(
                    JSONRPCErrorCode::INVALID_PARAMS,
                    "each plan must be an object",
                )
            })?;
            let name = obj.get("name").and_then(|v| v.as_str()).ok_or_else(|| {
                JSONRPCError::new(JSONRPCErrorCode::INVALID_PARAMS, "each plan needs name")
            })?;
            let room = obj.get("room").and_then(|v| v.as_str()).ok_or_else(|| {
                JSONRPCError::new(JSONRPCErrorCode::INVALID_PARAMS, "each plan needs room")
            })?;
            let code = obj
                .get("latticeCode")
                .and_then(|v| v.as_str())
                .ok_or_else(|| {
                    JSONRPCError::new(
                        JSONRPCErrorCode::INVALID_PARAMS,
                        "each plan needs latticeCode",
                    )
                })?;
            let model = obj
                .get("embeddingModelID")
                .and_then(|v| v.as_str())
                .ok_or_else(|| {
                    JSONRPCError::new(
                        JSONRPCErrorCode::INVALID_PARAMS,
                        "each plan needs embeddingModelID",
                    )
                })?;
            let sensitivity = decode_sensitivity(obj.get("sensitivity").and_then(|v| v.as_str()));
            Ok(PlanInput {
                name: name.to_owned(),
                room: room.to_owned(),
                lattice_code: code.to_owned(),
                embedding_model_id: model.to_owned(),
                sensitivity,
            })
        })
        .collect()
}

fn decode_sensitivity(name: Option<&str>) -> i64 {
    match name {
        Some("elevated") => AdjectiveSensitivity::Elevated.raw_value(),
        Some("restricted") => AdjectiveSensitivity::Restricted.raw_value(),
        Some("secret") => AdjectiveSensitivity::Secret.raw_value(),
        _ => AdjectiveSensitivity::Normal.raw_value(),
    }
}

/// Decode the recall filter for recipe tools. Mirrors Swift
/// `RecipeTools.decodeFilter(_:)` — defaults to Unconfirmed so freshly-
/// captured rows are visible.
fn decode_recipe_filter(args: &BTreeMap<String, JsonValue>) -> Filter {
    match args.get("filter").and_then(|v| v.as_str()) {
        Some("userConfirmed") => Filter::UserConfirmed,
        Some("exportable") => Filter::Exportable,
        Some("contained") => Filter::Contained,
        Some("currentlyBelieve") => Filter::CurrentlyBelieve,
        _ => Filter::Unconfirmed,
    }
}

/// Convert a `RecipeRunError` to a `JSONRPCError` for out-of-band recipe
/// failures. Recipe-level refusals (RecipeError) that are "expected" errors
/// should be converted to `error_result` by the caller instead — this is
/// for genuine substrate failures.
fn error_from_recipe(e: cognition_kit::RecipeRunError) -> JSONRPCError {
    JSONRPCError::new(JSONRPCErrorCode::TOOL_DISPATCH_FAILURE, format!("{e:?}"))
}
